flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMaster.java

```diff
@@ -699,13 +701,25 @@ public void declineCheckpoint(DeclineCheckpoint decline) {
 		final CheckpointCoordinator checkpointCoordinator = executionGraph.getCheckpointCoordinator();
 
 		if (checkpointCoordinator != null) {
-			getRpcService().execute(() -> {
+			CompletableFuture<Boolean> shouldFailJobFuture = getRpcService().execute(() -> {
 				try {
-					checkpointCoordinator.receiveDeclineMessage(decline);
+					return checkpointCoordinator.receiveDeclineMessage(decline);
 				} catch (Exception e) {
 					log.error("Error in CheckpointCoordinator while processing {}", decline, e);
 				}
+
+				return false;
 			});
+
+			try {
+				boolean shouldFailJob = shouldFailJobFuture.get(rpcTimeout.toMilliseconds(), TimeUnit.MILLISECONDS);
```

Comment:
I think this line is executed in the job master's main-thread and should therefore not be blocking. I think the it would be better to chain `failGlobal` after the `shouldFailJobFuture` completion to complete async in the job master main-thread.

Reply:
Good suggestion, accepted.